testsuite: improve in-tree MPI testing

garlick opened the pull request:
Our in-tree MPI testing is pretty minimal, basically just covering `MPI_Init()`, `MPI_Barrier()`, and `MPI_Finalize()`. This PR cleans up what we have, then adds `MPI_Abort()` coverage, and some basic comms tests pulled in from the MPICH source tree.

garlick commented:
Well that was a horror show in CI.  The `MPI_Abort()` tests were hanging and causing subsequent tests to not run because cores were still occupied.  I ran this down to a mismatch in the PMI-1 wire protocol between our implementation and the one in MPICH, so fixed that, and also moved the abort tests to their own sharness script.

garlick commented:
Added a check in the `MPI_Abort()` to skip the test if using mpich < 3.2.0 as described in #3277.

garlick commented:
This is probably OK for a review now.  There is a tiny bit of code change in here and it happens not to be covered, so the coverage is very low.  I don't think it's really worth concocting a PMI client that exercises this code as it's pretty obvious.

grondo commented:
I get a failure on my system (Ubuntu 18.04, mpich 3.3~a2-4)

Not exactly sure what to make of it though ...

```
grondo@asp:~/git/fc.git/t$ mpichversion
MPICH Version:    	3.3a2
MPICH Release date:	Sun Nov 13 09:12:11 MST 2016
MPICH Device:    	ch3:nemesis
```

```
expecting success:
	run_timeout 30 flux mini run -n 3 $OPTS \
		${FLUX_BUILD_DIR}/t/mpi/mpich_basic/adapt

0.039s: flux-shell[0]: DEBUG: 0: task_count=3 slot_count=3 cores_per_slot=1 slots_per_node=-1
0.040s: flux-shell[0]: DEBUG: 0: tasks [0-2] on cores 0-2
0.040s: flux-shell[0]: DEBUG: Loading /home/grondo/git/fc.git/src/shell/initrc.lua
0.044s: flux-shell[0]: DEBUG: output: batch timeout = 0.500s
0.045s: flux-shell[0]: DEBUG: affinity: disabling affinity due to cpu-affinity=off
2: asp
0: asp
1: asp
To determine 2 <-> 0       latency, using 262144 reps.
To determine       0 <-> 1 latency, using 262144 reps.
To determine 2 <-- 0 --> 1 latency, using 131072 reps
Latency20_ : 0.000000306
Latency_01 : 0.000000346
Latency201 : 0.000000651
Now starting main loop
  0:         0 bytes 262144 times -->    0.00 Mbps in 0.000000316 sec
  0:         0 bytes 262144 times -->    0.00 Mbps in 0.000000353 sec
  0:         0 bytes 131072 times -->  0.000000432 0.000000538 0.000000599 0.000000632 0.000000652 0.000000687 0.000000696 0.000000698 0.000000699 0.000000699 0.000000698 0.000000699 0.000000699 0.000000701 0.000000702 0.000000701 0.000000700 0.000000699
not ok 10 - ANL adapt works
#
#		run_timeout 30 flux mini run -n 3 $OPTS \
#			${FLUX_BUILD_DIR}/t/mpi/mpich_basic/adapt
#

# failed 1 among 10 test(s)
```

garlick commented:
Oops!  There shouldn't have been a 30s timeout on that test.  Just ran it to completion and it takes on the order of 30m on my laptop.

Fixed, rebased on master, and re-pushed.

Aside: I haven't looked in much detail at what those MPICH "basic" tests do.  My goal was just to make sure that there was not some simple MPI idiom that fails because Flux didn't provide something in the runtime environment.  Later perhaps we can refine these tests.

grondo commented:
If you'd like we can push an update to the `fluxrm/testenv:focal` image so you can drop a24c5a1

garlick commented:
> If you'd like we can push an update to the fluxrm/testenv:focal image so you can drop a24c5a1

That'd be great, thanks.  LMK when that's done and then I'll force push this without that commit, and we'll see if tests still pass.  (They should, I built a test docker image as helpfully explained in `src/test/docker/README.md`).

grondo commented:
flux-sched check failed due to
```
FAIL: t1017-rv1-bootstrap.t 4 - rv1-bootstrap: resource idempotency preserved
```

A known issue (flux-framework/flux-sched#799). I'll restart all checks.

grondo commented on t/t3003-mpi-abort.t:
```diff
@@ -0,0 +1,74 @@
+# get a PMI server trace in the CI output in case we need it to debug!
+test_expect_success 'MPI_Abort size=1 with PMI server tracing enabled' '
+	${FLUX_BUILD_DIR}/t/mpi/version | head -1 &&
+	! run_timeout 60 flux mini run $OPTS -overbose=2 \
+		${FLUX_BUILD_DIR}/t/mpi/abort 0
```
Use of `test_must_fail` is preferred over `!` in sharness tests (it can detect a segfault or other unexpected failure mode and diagnose it as such). In this case there is probably little difference, but it is good to avoid use of `!` so it doesn't get further propagated when tests are copied or used as a reference.

```suggestion
	test_must_fail run_timeout 60 flux mini run $OPTS -overbose=2 \
		${FLUX_BUILD_DIR}/t/mpi/abort 0
```

Same goes for the other uses of `!` below.

garlick replied:
Oh for some reason I didn't think that would work in combination with `run_timeout` but it does seem to work.  Will fix.

grondo reviewed (approved):
LGTM!

garlick commented:
Thanks!  I'm setting MWP before something else breaks.
(that doesn't inspire confidence does it :-)
